web/survey/models.py

```diff
@@ -428,91 +435,11 @@ class Meta:
-        is_tagged = False
-        for tsd in tsds:
-            if tsd.tag in release_tags:
-                is_tagged = True
-
-        if (settings.PROJECT in self.name) and is_tagged:
-            raise Exception(f"Cannot overwrite a released source {self.name}.")
-        super(Source, self).save(*args, **kwargs)
-
-    def delete(self, *args, **kwargs):
-        """Cannot delete released sources."""
-        release_tags = Tag.objects.filter(type='release')
-        sds = SourceDetection.objects.filter(source=self)
-        tsds = TagSourceDetection.objects.filter(source_detection__in=sds)
-        is_tagged = False
-        for tsd in tsds:
-            if tsd.tag in release_tags:
-                is_tagged = True
-        if (settings.PROJECT in self.name) and is_tagged:
-            raise Exception(f"Cannot overwrite a released source {self.name}.")
-        super(Source, self).delete(*args, **kwargs)
-
-    class Meta:
-        managed = False
-        db_table = 'source'
-        unique_together = (('name', ),)
-
-
-class SourceDetection(models.Model):
-    id = models.BigAutoField(primary_key=True)
-    source = models.ForeignKey(Source, on_delete=models.DO_NOTHING)
-    detection = models.OneToOneField(Detection, on_delete=models.DO_NOTHING)
-    added_at = models.DateTimeField(blank=True, null=True)
-
-    def save(self, *args, **kwargs):
-        """Do not save changes for released source detections."""
-        internal_release_tag = Tag.objects.get(name='Internal Release')
-        tsds = TagSourceDetection.objects.filter(source_detection=self)
-        is_released = False
-        if internal_release_tag in tsds:
-            is_released = True
-        if (settings.PROJECT in self.source.name) and is_released:
-            raise Exception(f"Cannot change source detection pointing to a released source {self.source.name}.")
-        super(SourceDetection, self).save(*args, **kwargs)
-
-    def delete(self, *args, **kwargs):
-        """Do not delete released sources."""
-        internal_release_tag = Tag.objects.get(name='Internal Release')
-        tsds = TagSourceDetection.objects.filter(source_detection=self)
-        is_released = False
-        if internal_release_tag in tsds:
-            is_released = True
-        if (settings.PROJECT in self.source.name) and is_released:
-            raise Exception(f"Cannot change source detection pointing to released source {self.source.name}.")
-        super(SourceDetection, self).delete(*args, **kwargs)
-
-    def __str__(self):
-        return f"{self.source}, {self.detection}"
-
-    class Meta:
-        managed = False
-        db_table = 'source_detection'
-
-
 class ExternalConflict(models.Model):
     id = models.BigAutoField(primary_key=True)
     run = models.ForeignKey(Run, on_delete=models.DO_NOTHING)
     detection = models.ForeignKey(Detection, on_delete=models.DO_NOTHING)
-    conflict_source_detection_ids = ArrayField(
+    conflict_detection_ids = ArrayField(
```

Comment:
It's a good idea to capture relationships between entities using related tables and associated keys. Using data types to capture relationships means the database engine can not enforce rules which may cause data integrity and consistency issues.

Reply:
This should be updated now and uses related tables for the run, detection and conflict. The code is:

class ExternalConflict(models.Model):
    id = models.BigAutoField(primary_key=True)
    run = models.ForeignKey(Run, on_delete=models.CASCADE)
    detection = models.ForeignKey(Detection, db_column='detection_id', related_name='detection', to_field='id', on_delete=models.CASCADE)
    conflict_detection = models.ForeignKey(Detection, db_column='conflict_detection_id', related_name='conflict_detection', to_field='id', on_delete=models.CASCADE)

    class Meta:
        managed = False
        db_table = 'external_conflict'